crates/re_log_types/src/data_table_batcher.rs

```diff
@@ -0,0 +1,787 @@
+    fn flush() -> (Self, Receiver<()>) {
+        let (tx, rx) = crossbeam::channel::bounded(0); // oneshot
+        (Self::Flush(tx), rx)
+    }
+}
+
+impl DataTableBatcher {
+    /// Creates a new [`DataTableBatcher`] using the passed in `config`.
+    ///
+    /// The returned object must be kept in scope: dropping it will trigger a clean shutdown of the
+    /// batcher.
+    #[must_use = "Batching threads will automatically shutdown when this object is dropped"]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn new(config: DataTableBatcherConfig) -> DataTableBatcherResult<Self> {
+        let (tx_cmds, rx_cmd) = match config.max_commands_in_flight {
+            Some(cap) => crossbeam::channel::bounded(cap as _),
+            None => crossbeam::channel::unbounded(),
+        };
+
+        let (tx_table, rx_tables) = match config.max_tables_in_flight {
+            Some(cap) => crossbeam::channel::bounded(cap as _),
+            None => crossbeam::channel::unbounded(),
+        };
+
+        let cmds_to_tables_handle = {
+            const NAME: &str = "DataTableBatcher::cmds_to_tables";
+            std::thread::Builder::new()
+                .name(NAME.into())
+                .spawn({
+                    let config = config.clone();
+                    move || batching_thread(config, rx_cmd, tx_table)
+                })
+                .map_err(|err| DataTableBatcherError::SpawnThread {
+                    name: NAME,
+                    err: Box::new(err),
+                })?
+        };
+
+        re_log::debug!(?config, "creating new table batcher");
+
+        let inner = DataTableBatcherInner {
+            tx_cmds,
+            rx_tables,
+            cmds_to_tables_handle: Some(cmds_to_tables_handle),
+        };
+
+        Ok(Self {
+            inner: Arc::new(inner),
+        })
+    }
+
+    // --- Send commands ---
+
+    /// Pushes a [`DataRow`] down the batching pipeline.
+    ///
+    /// This will call [`DataRow::compute_all_size_bytes`] from the batching thread!
+    ///
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    #[inline]
+    pub fn push_row(&self, row: DataRow) {
+        self.inner.push_row(row);
+    }
+
+    /// Initiates a flush of the pipeline and returns immediately.
+    ///
+    /// This does **not** wait for the flush to propagate (see [`Self::flush_blocking`]).
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    #[inline]
+    pub fn flush_async(&self) {
+        self.inner.flush_async();
+    }
+
+    /// Initiates a flush the batching pipeline and waits for it to propagate.
+    ///
+    /// See [`DataTableBatcher`] docs for ordering semantics and multithreading guarantees.
+    #[inline]
+    pub fn flush_blocking(&self) {
+        self.inner.flush_blocking();
+    }
+
+    // --- Subscribe to tables ---
+
+    /// Returns a _shared_ channel in which are sent the batched [`DataTable`]s.
```

Comment:
Any particular reason to make this shared? Is there value in a new receiver jumping in mid-stream? I worry about the usefulness of the data in that case.

Reply:
The channel is already mpmc by nature, so I didn't see the point of umping through extra hoops to turn it back into an mpsc one. Also parallel consumers might come in handy at some point.